extensions/extensionservice/org.eclipse.smarthome.extensionservice.marketplace/src/main/java/org/eclipse/smarthome/extensionservice/marketplace/internal/MarketplaceExtensionService.java

```diff
@@ -0,0 +1,404 @@
+                    break;
+            }
+        }
+    }
+
+    private void installRuleTemplate(String extensionId, MarketplaceExtension mpExt) {
+        String url = mpExt.getDownloadUrl();
+        try {
+            String template = getTemplate(url);
+            marketplaceRuleTemplateProvider.addTemplateAsJSON(extensionId, template);
+            postInstalledEvent(extensionId);
+        } catch (ParsingException e) {
+            postFailureEvent(extensionId, "Template is not valid.");
+            logger.error("Rule template from marketplace is invalid: {}", e.getMessage());
+        } catch (IOException e) {
+            postFailureEvent(extensionId, "Template cannot be downloaded.");
+            logger.error("Rule template from marketplace cannot be downloaded: {}", e.getMessage());
+        }
+    }
+
+    private void installBinding(String extensionId, MarketplaceExtension mpExt) {
+        String url = mpExt.getDownloadUrl();
+        try {
+            Bundle bundle = bundleContext.installBundle(url);
+            installedBindings.put(extensionId, bundle.getBundleId());
+            persistInstalledBindingsMap(installedBindings);
+            postInstalledEvent(extensionId);
+        } catch (BundleException e) {
+            postFailureEvent(extensionId, "Binding cannot be installed: " + e.getMessage());
+        }
+    }
+
+    private void uninstallBinding(String extensionId) {
+        Long id = installedBindings.get(extensionId);
+        if (id != null) {
+            Bundle bundle = bundleContext.getBundle(id);
+            if (bundle != null) {
+                try {
+                    bundle.uninstall();
+                    installedBindings.remove(extensionId);
+                    persistInstalledBindingsMap(installedBindings);
+                    postUninstalledEvent(extensionId);
+                } catch (BundleException e) {
+                    postFailureEvent(extensionId, "Failed deinstalling binding: " + e.getMessage());
+                }
+            } else {
+                // we do not have such a bundle, so let's remove it from our internal map
+                installedBindings.remove(extensionId);
+                persistInstalledBindingsMap(installedBindings);
+                postFailureEvent(extensionId, "Id not known.");
+            }
+        } else {
+            postFailureEvent(extensionId, "Id not known.");
+        }
+    }
+
+    private String getTemplate(String urlString) throws IOException {
+        URL url = new URL(urlString);
+        return IOUtils.toString(url);
+    }
+
+    private void postInstalledEvent(String extensionId) {
+        if (eventPublisher != null) {
+            Event event = ExtensionEventFactory.createExtensionInstalledEvent(extensionId);
+            eventPublisher.post(event);
+        }
+    }
+
+    private void postUninstalledEvent(String extensionId) {
+        if (eventPublisher != null) {
```

Comment:
I assume this function must only be called if the service is activated.
EventPublisher is a mandatory reference, so no need to check for nullness.